File: ui/src/app/services/auth.service.ts
/*
 * Copyright 2023 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { UserService } from "@appServices/user.service";
import { concatMap } from 'rxjs/operators';

@Injectable({
  providedIn: 'root',
})
export class AuthService {

  private serviceUrl = '/api/auth';  // URL to web api

  httpOptions = {
    headers: new HttpHeaders({
      'Content-Type': 'application/x-www-form-urlencoded',
      'Accept': 'application/json'
    })
  };

  constructor(private router: Router,
              private http: HttpClient,
              private userService: UserService) {
  }

  /**
   * Present the login credentials to the authentication service on the server.
   * It will return the username if the credentials match a user, or {@code null} if not.
   * Then request the user profile from the user service.
   *
   * @param username the username.
   * @param password the password.
   * @return an <Observable<User | null> for the User profile logged in, or {@code null} if the
   *         login failed.
   */
  login(username: string, password: string) {
    return this.http.post<string>(`${this.serviceUrl}/login`,
      `username=${username}&password=${password}`, this.httpOptions)
      .pipe(concatMap(username => this.userService.findByUsername(username))
      );
  }
  //
  // /**
  //  * Log the user out. Return to the login screen.
  //  */
  // logout() {
  //   // remove user from local storage to log user out
  //   sessionStorage.removeItem('user');
  //   this.userSubject.next(null);
  //   this.router.navigate(['/login']);
  // }
}

File: service/src/main/java/org/kathrynhuxtable/heroes/service/AuthService.java
/*
 * Copyright 2002-2018 the original author or authors.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package org.kathrynhuxtable.heroes.service;

import java.util.ArrayList;
import java.util.List;

import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import org.kathrynhuxtable.heroes.service.bean.AuthInfo;
import org.kathrynhuxtable.heroes.service.bean.User;

@Slf4j
@Component
public class AuthService {

	private List<AuthInfo> users;

	@Autowired
	private SecurityService securityService;

	public AuthService() {
		users = new ArrayList<>();

		users.add(makeUser(1, "admin", "secret1"));
		users.add(makeUser(2, "user", "pass1"));
	}

	public User authenticate(String username, String password) {
		for (AuthInfo user : users) {
			if (user.getUsername().equals(username) && user.getPassword().equals(password)) {
				return findUser(user.getUsername());
			}
		}
		return null;
	}

	public User findUser(String username) {
		for (AuthInfo user : users) {
			if (user.getUsername().equalsIgnoreCase(username)) {
				return securityService.findUser(user.getId());
			}
		}
		return null;
	}

	private AuthInfo makeUser(int id, String username, String password) {
		return AuthInfo.builder().id(id).username(username).password(password).build();
	}
}
